Team6220_PowerPlay/src/main/java/org/firstinspires/ftc/team6220_PowerPlay/BaseOpMode.java

```diff
@@ -127,37 +157,97 @@ public void driveWithIMU(double xPower, double yPower, double tPower) {
-        } else {
-            servoGrabber.setPosition(Constants.GRABBER_CLOSE_POSITION);
-        }
+    public void driveGrabber(double position) {
+        servoGrabber.setPosition(position);
     }
 
     /**
-     * this method will allow the slides to move upwards and downwards given a specified target position
-     * @param targetPosition target position for vertical slides motors in ticks
+     * this method will allow the slides to move to a specified target position
+     * @param targetPosition target position for slides motors in ticks
      */
     public void driveSlides(int targetPosition) {
+        int error = targetPosition - motorLeftSlides.getCurrentPosition();
+        double motorPower = error * Constants.SLIDE_MOTOR_KP;
+
+        // slides not yet at target position
+        if (Math.abs(error) > Constants.ROBOT_SLIDE_TOLERANCE_TICKS) {
+            // slides going down - joystick
+            if (error < 0 && error > -200) {
+                motorLeftSlides.setPower(-0.3);
+                motorRightSlides.setPower(-0.3);
+                // slides going down - bumpers
+            } else if (error < -200) {
+                motorLeftSlides.setPower(-1.0);
+                motorRightSlides.setPower(-1.0);
+                // slides going up - proportional control
+            } else {
+                motorLeftSlides.setPower(motorPower);
+                motorRightSlides.setPower(motorPower);
+            }
+            // slides at target position
+        } else {
+            motorLeftSlides.setPower(Constants.SLIDE_FEEDFORWARD);
+            motorRightSlides.setPower(Constants.SLIDE_FEEDFORWARD);
+        }
+    }
+
+    // centers the cone on the junction top
+    public void centerJunctionTop(GrabberCameraPipeline pipeline) {
+        double xOffset, yOffset;
+
+        do {
+            xOffset = pipeline.xPosition - Constants.CAMERA_CENTER_X;
+            yOffset = Constants.CAMERA_CENTER_Y - pipeline.yPosition;
 
+            // center the cone on the junction top
+            if (pipeline.detected) {
+                driveWithIMU(Constants.JUNCTION_TOP_CENTERING_KP * Math.signum(xOffset), Constants.JUNCTION_TOP_CENTERING_KP * Math.signum(yOffset), 0.0);
+            } else {
+                break;
+            }
+
+        // while the cone isn't centered over the junction
+        } while (Math.abs(xOffset) > Constants.JUNCTION_TOP_TOLERANCE || Math.abs(yOffset) > Constants.JUNCTION_TOP_TOLERANCE);
+
+        stopDriveMotors();
     }
 
-    /**
-     * this method will allow the turntable to turn clockwise or counterclockwise given a specified power and position
-     * @param position target position of turntable motor in ticks
-     */
-    public void driveTurntable(int position) {
-        motorTurntable.setTargetPosition(position);
+    // moves towards the cone stack while centering on it until the stack fills the entire camera view
```

Comment:
All methods should have a javadoc comment

Reply:
Made javadoc comment